compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3411,20 +3411,23 @@ class Typer extends Namer
         else if (canTryGADTHealing) {
           // try recovering with a GADT approximation
           val nestedCtx = ctx.fresh.setNewTyperState()
-          val res =
-            readapt(
-              tree = tpd.Typed(tree, TypeTree(Inferencing.approximateGADT(wtp))),
-              shouldTryGadtHealing = false,
-            )(using nestedCtx)
-          if (!nestedCtx.reporter.hasErrors) {
-            // GADT recovery successful
-            nestedCtx.typerState.commit()
-            res
-          } else {
-            // otherwise fail with the error that would have been reported without the GADT recovery
-            err.typeMismatch(tree, pt, failure)
+          val approximated = Inferencing.approximateGADT(wtp)(using nestedCtx)
+          if (tree.tpe <:< approximated) {
```

Comment:
Shouldn't that check be done in the nestedCtx?

Reply:
That's sadly not the part that's making it explode. Not even making the check frozen and doing it in nestedCtx helps. I have no idea why, seems that we will need to take a more scorched-earth type of approach.